middleware/throttle.go

```diff
@@ -34,21 +54,46 @@ type token struct{}
 
 // throttler limits number of currently processed requests at a time.
 type throttle struct {
-	h      chi.Handler
-	tokens chan token
+	h             chi.Handler
+	tokens        chan token
+	backlogtokens chan token
+	timeout       time.Duration
 }
 
 // TODO: add support for a backlog
 
 // ServeHTTPC implements chi.Handler interface.
 func (t *throttle) ServeHTTPC(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	timer := time.NewTimer(t.timeout)
+
 	select {
+	case <-timer.C:
+		httpStatus(w, http.StatusServiceUnavailable)
+		return
 	case <-ctx.Done():
+		httpStatus(w, http.StatusServiceUnavailable)
 		return
-	case tok := <-t.tokens:
+	case btok := <-t.backlogtokens:
 		defer func() {
-			t.tokens <- tok
+			t.backlogtokens <- btok
 		}()
-		t.h.ServeHTTPC(ctx, w, r)
+		select {
+		case <-timer.C:
+			httpStatus(w, http.StatusGatewayTimeout)
+			return
+		case <-ctx.Done():
```

Comment:
There are several places such like this. I think that all timeouts should trigger http.StatusGatewayTimeout instead of http.StatusServiceUnavailable -- objections?